Root nodes for the duration of their CSS transitions

jdm opened the pull request:
This ensures that we can pass a node address as part of the asynchronous
transition end notification, making it safe to fire the corresponding
DOM event on the node from the script thread. Without explicitly rooting
this node when the transition starts, we risk the node being GCed before
the transition is complete.

---
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #14972
- [X] There are tests for these changes

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16295)
<!-- Reviewable:end -->

jdm committed:
Root nodes for the duration of their CSS transitions.

This ensures that we can pass a node address as part of the asynchronous
transition end notification, making it safe to fire the corresponding
DOM event on the node from the script thread. Without explicitly rooting
this node when the transition starts, we risk the node being GCed before
the transition is complete.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/matching.rs, components/style/animation.rs
 * @kichjang: components/script/script_thread.rs, components/script/dom/window.rs, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script_layout_interface/rpc.rs
 * @fitzgen: components/script/script_thread.rs, components/script/dom/window.rs, components/script_traits/lib.rs, components/script_traits/lib.rs, components/script_layout_interface/rpc.rs
 * @emilio: components/layout/animation.rs, components/layout/query.rs, components/style/matching.rs, components/style/animation.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 66d93d2c7f42e2149e775875f10103b48f94d025 with merge 8b5c8b8f4a611ca92429c9f9ba2427e0caa4b874...

bors-servo commented:
:broken_heart: Test failed - [android](http://build.servo.org/builders/android/builds/6178)

jdm commented:
As I feared, there could be problems with this approach when the animations are triggered from the compositor:
```
  ▶ CRASH [expected TIMEOUT] /dom/events/EventListener-invoke-legacy.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ assertion failed: rw_data.newly_transitioning_nodes.is_empty() (thread LayoutThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /Users/servo/buildbot/slave/mac-rel-wpt1/build/components/layout_thread/lib.rs:1411)
  │ 2017-04-07 01:35:16.662 servo[70066:646481199] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x10459e49e - backtrace::backtrace::trace::h47a32bf8c6f00c9b
  │    1:        0x10459ebec - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:        0x102e7d5b8 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::hab3ba3e402ac1f0b
  │    3:        0x1049ff9c9 - std::panicking::rust_panic_with_hook::h05996066754c6be9
  │    4:        0x102faa4f4 - std::panicking::begin_panic::haab64c72e0aa1aab
  │    5:        0x10300ecdc - layout_thread::LayoutThread::perform_post_style_recalc_layout_passes::h139d5d3cd94b81e4
  │    6:        0x10300d2d5 - layout_thread::LayoutThread::tick_all_animations::h90b400a44eeb0e6f
  │    7:        0x103003306 - layout_thread::LayoutThread::handle_request_helper::h307a64d67c3c4130
  │    8:        0x102f944a0 - profile_traits::mem::ProfilerChan::run_with_memory_reporting::h6184f9f4fcb8e847
  │    9:        0x103001173 - _$LT$layout_thread..LayoutThread$u20$as$u20$layout_traits..LayoutThreadFactory$GT$::create::_$u7b$$u7b$closure$u7d$$u7d$::h8028856a97c8c244
  │   10:        0x102faadfd - std::panicking::try::do_call::ha664f0ed882ba9a8
  │   11:        0x104a0099a - __rust_maybe_catch_panic
  │   12:        0x102fc41e6 - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hc02c54adca1db771
  │   13:        0x1049fc4e4 - std::sys::imp::thread::Thread::new::thread_start::h4008e1859fbd98b8
  │   14:     0x7fff8af9f059 - _pthread_body
  │   15:     0x7fff8af9efd6 - _pthread_start
  │ ERROR:servo: assertion failed: rw_data.newly_transitioning_nodes.is_empty()
  │ Pipeline failed in hard-fail mode.  Crashing!
  └ thread panicked while processing panic. aborting.
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16316) made this pull request unmergeable. Please resolve the merge conflicts.

jdm committed:
Accumulate transitioning nodes inside the layout context.

jdm committed:
Send information to script as part of finishing layout.

This avoids the need for multiple layout RPC operations immediately
following return of control to script. This means that layout and
script can continue to operate in parallel at this point, rather
than one potentially waiting on the shared mutex to be unlocked.

jdm committed:
Warn about ignored transitions which can't be rooted.

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 226e43398367f5d7e3d4ce1e2513f79d7c9f098c with merge bef0a38c632440dd14f18e87968d5400409ed1a9...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3333)

jdm commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit 212f9984d7c08279671371c3b8121c5ad55f5172 with merge 39f1fc5b3b6b672b0738dd3f3036e71d00982e13...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6229), [arm32](http://build.servo.org/builders/arm32/builds/6448), [arm64](http://build.servo.org/builders/arm64/builds/6481), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6582), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3314), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3340), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6586), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6618), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3272), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3343), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1758)
State: approved= try=True

jdm commented:
@nox do you want to review this?

nox commented:
r? @nox

nox commented on components/script/dom/window.rs:
```diff
@@ -1244,6 +1245,9 @@ impl Window {
             }
         }
 
+        let newly_transitioning_nodes = self.layout_rpc.newly_transitioning_nodes();
```
Layout can send back garbage addresses and then everything goes to hell.

nox commented on components/script/dom/window.rs:
```diff
@@ -1244,6 +1245,9 @@ impl Window {
         }
 
+        let newly_transitioning_nodes = self.layout_rpc.newly_transitioning_nodes();
+        ScriptThread::note_newly_transitioning_nodes(newly_transitioning_nodes);
```
This function definitely should be unsafe.

nox commented on components/script/script_thread.rs:
```diff
@@ -574,6 +576,17 @@ impl ScriptThreadFactory for ScriptThread {
+            let js_runtime = script_thread.js_runtime.rt();
+            let new_nodes = nodes
+                .into_iter()
+                .map(|n| JS::from_ref(&*from_untrusted_node_address(js_runtime, n)));
```
Function `from_untrusted_node_address` should have definitely be `unsafe`.

nox commented on components/script/script_thread.rs:
```diff
@@ -1600,11 +1615,27 @@ impl ScriptThread {
-        let node = unsafe { ServoLayoutNode::from_unsafe(&unsafe_node) };
-        let node = unsafe { node.get_jsmanaged().get_for_script() };
-        let window = window_from_node(node);
+    fn handle_transition_event(&self, unsafe_node: UntrustedNodeAddress, name: String, duration: f64) {
```
This should be unsafe.

nox commented on components/layout/animation.rs:
```diff
@@ -86,8 +89,8 @@ pub fn update_animation_state(constellation_chan: &IpcSender<ConstellationMsg>,
-            if let Animation::Transition(_, unsafe_node, _, ref frame, _) = running_animation {
-                script_chan.send(ConstellationControlMsg::TransitionEnd(unsafe_node,
+            if let Animation::Transition(node, _, ref frame, _) = running_animation {
+                script_chan.send(ConstellationControlMsg::TransitionEnd(node.to_untrusted_node_address(),
```
Why can't we send back the `node` itself?

jdm replied:
It's a layout type.

nox commented on components/script/script_thread.rs:
```diff
@@ -574,6 +576,17 @@ impl ScriptThreadFactory for ScriptThread {
 }
 
 impl ScriptThread {
+    pub fn note_newly_transitioning_nodes(nodes: Vec<UntrustedNodeAddress>) {
```
Can GC happen before this even gets a change to be called?

IMO this whole function should be unsafe.

jdm replied:
Anything that has a transition starting during layout is part of the DOM, therefore cannot be GCed. There's no need for this to be unsafe.

nox replied:
It needs to be unsafe.

```thread
ScriptThread::note_newly_transitioning_nodes(vec![UntrustedNodeAddress:from_id(0xdeadbeaf))
```

nox reviewed (changes requested):
This does not look as safe as I would like to be but this is certainly better than the code code we currently have.

That being said, I don't think we will ever have sound transition code for as long as we continue to send `UnsafeNodeAddress`-like types between layout and script.

jdm committed:
Make methods storing layout-originating nodes unsafe.

jdm commented:
Addressed comments about methods that should be unsafe.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16802) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented on components/script/dom/node.rs:
```diff
@@ -927,20 +927,18 @@ fn first_node_not_in<I>(mut nodes: I, not_in: &[NodeOrString]) -> Option<Root<No
-        // https://github.com/servo/servo/issues/6383
-        let candidate: uintptr_t = mem::transmute(candidate.0);
+    // https://github.com/servo/servo/issues/6383
+    let candidate: uintptr_t = mem::transmute(candidate.0);
```
Please don't use transmute.

nox commented on components/script/dom/node.rs:
```diff
@@ -927,20 +927,18 @@ fn first_node_not_in<I>(mut nodes: I, not_in: &[NodeOrString]) -> Option<Root<No
-        }
-        let boxed_node = conversions::private_from_object(object) as *const Node;
-        Root::from_ref(&*boxed_node)
+    let object: *mut JSObject = mem::transmute(candidate);
```
Please don't use transmute.

nox reviewed (approved):
Looks good, just two nits, feel free to not address them if you don't care.

jdm commented:
@bors-servo: r=nox

bors-servo commented:
:pushpin: Commit d26e45b has been approved by `nox`

<!-- @bors-servo r=nox d26e45b2f6c4ddf97b3c38ee2eb16bf81367b0a8 -->

jdm commented:
@bors-servo: r=nox

bors-servo commented:
:pushpin: Commit b0bf2b4 has been approved by `nox`

<!-- @bors-servo r=nox b0bf2b4bad636acfba66d55571b417ebae795408 -->

bors-servo commented:
:hourglass: Testing commit b0bf2b4bad636acfba66d55571b417ebae795408 with merge fa251ec96b445b9ba8439d76e05870a88c2caa0f...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/6945), [arm32](http://build.servo.org/builders/arm32/builds/7154), [arm64](http://build.servo.org/builders/arm64/builds/7193), [linux-dev](http://build.servo.org/builders/linux-dev/builds/7293), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4040), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4052), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/7270), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/7319), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3970), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4050), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/2499)
Approved by: nox
Pushing fa251ec96b445b9ba8439d76e05870a88c2caa0f to master...
